Display custom metrics for pods if they are available.

mwringe opened the pull request:
The Hawkular OpenShift Agent is capable of gathering a few custom metrics from pods which are deployed.

This PR will allow these pods to be displayed under the pod's metric page

spadgett commented on app/scripts/directives/podMetrics.js:
```diff
@@ -95,6 +95,38 @@ angular.module('openshiftConsole')
+            angular.forEach(response, function(metric) {
+
+              // set the label to the description if specified
+              var label = (metric.description !== undefined) ? metric.description : metric.name;
```
```js
var label = metric.description || metric.name;
```

spadgett commented on app/scripts/directives/podMetrics.js:
```diff
@@ -95,6 +95,38 @@ angular.module('openshiftConsole')
+              var label = (metric.description !== undefined) ? metric.description : metric.name;
+              
+              // get the unit value if specified
+              var unit = (metric.unit !== undefined) ? metric.unit : "unknown";
```
```js
var unit = metric.unit || "unknown";
```

But maybe the empty string is better if no unit set?

```js
var unit = metric.unit || "";
```

spadgett commented on app/scripts/services/metrics.js:
```diff
@@ -169,6 +169,46 @@ angular.module("openshiftConsole")
+            };
+            metrics.push(metric);
+          });
+           return metrics;
```
indentation

spadgett commented on app/scripts/services/metrics.js:
```diff
@@ -169,6 +169,46 @@ angular.module("openshiftConsole")
+              description: value.tags.description,
+            };
+            metrics.push(metric);
+          });
```
Suggest using `_.map`. (Untested, but I think this will work.)

```js
return _.map(response.data, function(value) {
  return {
    name: value.tags.metric_name,
    unit: value.tags.units,
    description: value.tags.description
  };
});
```

spadgett commented on app/scripts/services/metrics.js:
```diff
@@ -169,6 +169,46 @@ angular.module("openshiftConsole")
+            var metric = {
+              name: value.tags.metric_name,
+              unit: value.tags.units,
+              description: value.tags.description,
```
Extra trailing comma

spadgett commented:
cc @jwforres @derekwaynecarr @DirectXMan12

mwringe commented:
@spadgett Ok, I made the requested changes to the code.

Custom metrics is something we want for 1.5, we are hoping to have something in the console to display these metrics. If we can not get it the console at this point in time, we will look to create a custom console extension or rely on something like grafana to display them.

spadgett commented:
@mwringe can you add a flag to disable custom metrics to constants.js? something like

https://github.com/openshift/origin-web-console/blob/master/app/scripts/constants.js#L66

mwringe commented:
@spadgett do we just want a flag to be added and the rest kept the same? Or a flag to be added and the drop down to still be used instead?

spadgett commented:
my opinion: flag added and the rest kept the same.

@jwforres let me know if you disagree

jwforres commented:
yeah based on @spadgett 's assessment about the risk of introducing the
dropdown at this point in the release I agree

On Mon, Jan 16, 2017 at 3:06 PM, Sam Padgett <notifications@github.com>
wrote:

> my opinion: flag added and the rest kept the same.
>
> @jwforres <https://github.com/jwforres> let me know if you disagree
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin-web-console/pull/1109#issuecomment-272951369>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABZk7ZGRK1L6-XNGqutortfMiI_RFv5Bks5rS83fgaJpZM4LiRhu>
> .
>

mwringe commented:
Update to include a flag to enable/disable custom metrics.

I also fixed an issue I found where we were not properly fetching counter based metrics and fixed a potential issue with encoding the metric name for the chart prefix.

spadgett commented on app/scripts/directives/podMetrics.js:
```diff
@@ -95,6 +95,45 @@ angular.module('openshiftConsole')
+
+                // note: the prefix needs to follow selector rules
+                // converting to base64 and cleaning up the special /,+,= characters
+                var prefix = btoa(metric.name).replace(/\+/g,"_62_").replace(/\//g,"_63_").replace(/\=/g,"");
```
Suggest this instead for the prefix:

```js
var prefix = _.uniqueId('custom-metric-');
```

mwringe replied:
awesome, that will work

spadgett commented on app/scripts/services/metrics.js:
```diff
@@ -50,7 +51,12 @@ angular.module("openshiftConsole")
       return getMetricsURL().then(function(metricsURL) {
-        var template = metricsURL + POD_GAUGE_TEMPLATE;
+        var template;
+        if (config.type !== undefined && config.type === "counter") {
```
Maybe just

```js
if (config.type === 'counter') { ... }
```

although probably worth a comment saying gauge is assumed when type is not set

mwringe replied:
ok

spadgett commented on app/scripts/services/metrics.js:
```diff
@@ -169,6 +175,44 @@ angular.module("openshiftConsole")
+        var url = metricsURL + "/metrics";
+
+        var params = {
+          tags: "custom_metric:true,pod_name:" + podName
```
Missed this on first review, but should we use pod UID instead in case a pod with the same name is created twice? (For pods that don't have generated names.)

mwringe replied:
oh, good catch. Thanks. I will fix that.

mwringe commented:
Ok, updated with the requested changes.

spadgett commented:
[merge]
